### PR TITLE
Create custom machineset for creating nodes with different instance type and node scale

### PR DESCRIPTION
**Pull request by ramkiperiy**
Create custom machineset for creating nodes with different instance type.
Using this machineset we can scale nodes for app pod creation/deletion.
This will not affect the OCS worker node.

ocs_ci/ocs/constants.py --> Updated with new yaml file path
ocs_ci/ocs/machine.py --> Added function to create and delete machineset
ocs_ci/templates/openshift-infra/machine-set.yaml --> machineset yaml file

Signed-off-by: Ramakrishnan <rperiyas@redhat.com>

### CODE REVIEW

**Comment by ramkiperiy** [comment]
Local run log output for using function create_custom_machineset and delete_custom_machineset.
Script used for testing present in pastebin http://pastebin.test.redhat.com/839554
`11:39:05 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-machine-api --kubeconfig /home/rperiyas/26thFeb/auth/kubeconfig get machinesets  -n openshift-machine-api -o yaml
11:39:06 - MainThread - root - INFO - Print yaml {'apiVersion': 'machine.openshift.io/v1beta1', 'kind': 'MachineSet', 'metadata': {'labels': {'machine.openshift.io/cluster-api-cluster': 'rperiyas-cls-w76c9'}, 'name': 'rperiyas-cls-w76c9-app-us-east-2c', 'namespace': 'openshift-machine-api'}, 'spec': {'replicas': 1, 'selector': {'matchLabels': {'machine.openshift.io/cluster-api-cluster': 'rperiyas-cls-w76c9', 'machine.openshift.io/cluster-api-machineset': 'rperiyas-cls-w76c9-app-us-east-2c'}}, 'template': {'metadata': {'labels': {'machine.openshift.io/cluster-api-cluster': 'rperiyas-cls-w76c9', 'machine.openshift.io/cluster-api-machine-role': 'app', 'machine.openshift.io/cluster-api-machine-type': 'app', 'machine.openshift.io/cluster-api-machineset': 'rperiyas-cls-w76c9-app-us-east-2c'}}, 'spec': {'metadata': {'labels': {'node-role.kubernetes.io/app': 'app-scale'}}, 'providerSpec': {'value': {'ami': {'id': 'ami-04a79d8d7cfa540cc'}, 'apiVersion': 'awsproviderconfig.openshift.io/v1beta1', 'blockDevices': [{'ebs': {'iops': 0, 'volumeSize': 120, 'volumeType': 'gp2'}}], 'credentialsSecret': {'name': 'aws-cloud-credentials'}, 'deviceIndex': 0, 'iamInstanceProfile': {'id': 'rperiyas-cls-w76c9-worker-profile'}, 'instanceType': 'm4.xlarge', 'kind': 'AWSMachineProviderConfig', 'placement': {'availabilityZone': 'us-east-2c', 'region': 'us-east-2'}, 'securityGroups': [{'filters': [{'name': 'tag:Name', 'values': ['rperiyas-cls-w76c9-worker-sg']}]}], 'subnet': {'filters': [{'name': 'tag:Name', 'values': ['rperiyas-cls-w76c9-private-us-east-2c']}]}, 'tags': [{'name': 'kubernetes.io/cluster/rperiyas-cls-w76c9', 'value': 'owned'}], 'userDataSecret': {'name': 'worker-user-data'}}}}}}}
11:39:06 - MainThread - ocs_ci.ocs.resources.ocs - INFO - Adding MachineSet with name rperiyas-cls-w76c9-app-us-east-2c
11:39:06 - MainThread - ocs_ci.utility.templating - INFO - apiVersion: machine.openshift.io/v1beta1
                                                           kind: MachineSet
                                                           metadata:
                                                             labels:
                                                               machine.openshift.io/cluster-api-cluster: rperiyas-cls-w76c9
                                                             name: rperiyas-cls-w76c9-app-us-east-2c
                                                             namespace: openshift-machine-api
                                                           spec:
                                                             replicas: 1
                                                             selector:
                                                               matchLabels:
                                                                 machine.openshift.io/cluster-api-cluster: rperiyas-cls-w76c9
                                                                 machine.openshift.io/cluster-api-machineset: rperiyas-cls-w76c9-app-us-east-2c
                                                             template:
                                                               metadata:
                                                                 labels:
                                                                   machine.openshift.io/cluster-api-cluster: rperiyas-cls-w76c9
                                                                   machine.openshift.io/cluster-api-machine-role: app
                                                                   machine.openshift.io/cluster-api-machine-type: app
                                                                   machine.openshift.io/cluster-api-machineset: rperiyas-cls-w76c9-app-us-east-2c
                                                               spec:
                                                                 metadata:
                                                                   labels:
                                                                     node-role.kubernetes.io/app: app-scale
                                                                 providerSpec:
                                                                   value:
                                                                     ami:
                                                                       id: ami-04a79d8d7cfa540cc
                                                                     apiVersion: awsproviderconfig.openshift.io/v1beta1
                                                                     blockDevices:
                                                                     - ebs:
                                                                         iops: 0
                                                                         volumeSize: 120
                                                                         volumeType: gp2
                                                                     credentialsSecret:
                                                                       name: aws-cloud-credentials
                                                                     deviceIndex: 0
                                                                     iamInstanceProfile:
                                                                       id: rperiyas-cls-w76c9-worker-profile
                                                                     instanceType: m4.xlarge
                                                                     kind: AWSMachineProviderConfig
                                                                     placement:
                                                                       availabilityZone: us-east-2c
                                                                       region: us-east-2
                                                                     securityGroups:
                                                                     - filters:
                                                                       - name: tag:Name
                                                                         values:
                                                                         - rperiyas-cls-w76c9-worker-sg
                                                                     subnet:
                                                                       filters:
                                                                       - name: tag:Name
                                                                         values:
                                                                         - rperiyas-cls-w76c9-private-us-east-2c
                                                                     tags:
                                                                     - name: kubernetes.io/cluster/rperiyas-cls-w76c9
                                                                       value: owned
                                                                     userDataSecret:
                                                                       name: worker-user-data
11:39:06 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-machine-api --kubeconfig /home/rperiyas/26thFeb/auth/kubeconfig create -f /tmp/MachineSetksnqn3rg -o yaml
11:39:08 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-machine-api --kubeconfig /home/rperiyas/26thFeb/auth/kubeconfig get MachineSet rperiyas-cls-w76c9-app-us-east-2c -n openshift-machine-api -o yaml
PASSED                                                                                                                                                                [100%]

============================================================================ 1 passed in 13.05s =============================================================================`

**Comment by ramkiperiy** [comment]
Output from OCP 
[rperiyas@localhost Pod]$ oc get machineset -n openshift-machine-api
NAME                                   DESIRED   CURRENT   READY   AVAILABLE   AGE
rperiyas-cls-w76c9-app-us-east-2a      1         1         1       1           10m
rperiyas-cls-w76c9-app-us-east-2b      1         1         1       1           14m
rperiyas-cls-w76c9-worker-us-east-2a   1         1         1       1           144m
rperiyas-cls-w76c9-worker-us-east-2b   1         1         1       1           144m
rperiyas-cls-w76c9-worker-us-east-2c   1         1         1       1           144m
[rperiyas@localhost Pod]$ oc get nodes
NAME                                         STATUS   ROLES        AGE     VERSION
ip-10-0-135-205.us-east-2.compute.internal   Ready    worker       135m    v1.16.2
ip-10-0-135-57.us-east-2.compute.internal    Ready    app,worker   6m51s   v1.16.2
ip-10-0-139-73.us-east-2.compute.internal    Ready    master       144m    v1.16.2
ip-10-0-144-248.us-east-2.compute.internal   Ready    master       145m    v1.16.2
ip-10-0-145-98.us-east-2.compute.internal    Ready    worker       135m    v1.16.2
ip-10-0-158-93.us-east-2.compute.internal    Ready    app,worker   11m     v1.16.2
ip-10-0-168-123.us-east-2.compute.internal   Ready    worker       135m    v1.16.2
ip-10-0-175-186.us-east-2.compute.internal   Ready    master       144m    v1.16.2
[rperiyas@localhost Pod]$ oc get machineset -n openshift-machine-api
NAME                                   DESIRED   CURRENT   READY   AVAILABLE   AGE
rperiyas-cls-w76c9-app-us-east-2a      1         1         1       1           11m
rperiyas-cls-w76c9-worker-us-east-2a   1         1         1       1           145m
rperiyas-cls-w76c9-worker-us-east-2b   1         1         1       1           145m
rperiyas-cls-w76c9-worker-us-east-2c   1         1         1       1           145m
[rperiyas@localhost Pod]$ oc get nodes
NAME                                         STATUS   ROLES        AGE    VERSION
ip-10-0-135-205.us-east-2.compute.internal   Ready    worker       136m   v1.16.2
ip-10-0-135-57.us-east-2.compute.internal    Ready    app,worker   8m     v1.16.2
ip-10-0-139-73.us-east-2.compute.internal    Ready    master       145m   v1.16.2
ip-10-0-144-248.us-east-2.compute.internal   Ready    master       146m   v1.16.2
ip-10-0-145-98.us-east-2.compute.internal    Ready    worker       136m   v1.16.2
ip-10-0-168-123.us-east-2.compute.internal   Ready    worker       136m   v1.16.2
ip-10-0-175-186.us-east-2.compute.internal   Ready    master       145m   v1.16.2
[rperiyas@localhost Pod]$

File: ocs_ci/ocs/machine.py [hunk 1]
```diff
@@ -138,6 +140,113 @@ def delete_machine_and_check_state_of_new_spinned_machine(machine_name):
...
+                    'subnet'
+                ]['filters'][0]['values'][0] = f"{cls_id}-private-{aws_zone}"
+                machineset_yaml['spec']['template']['spec']['providerSpec']['value'][
+                    'tags'
+                ][0]['name'] = f"kubernetes.io/cluster/{cls_id}"
+                logging.info(f"Print yaml {machineset_yaml}")
+
+                # Create new custom machineset
+                ms_obj = OCS(**machineset_yaml)
+                ms_obj.create()
```
Comment (keesturam) [inline ocs_ci/ocs/machine.py]: Please add validation to see if the machine set is actually created in the cluster.

File: ocs_ci/ocs/machine.py [hunk 2]
```diff
@@ -138,6 +140,113 @@ def delete_machine_and_check_state_of_new_spinned_machine(machine_name):
...
+        machine_set (str): Name of the machine set to be deleted
+        WARN: Make sure it's not OCS worker node machines set, if so then
+              OCS worker nodes and machine set will be deleted.
+
+    Returns:
+        bool: True if commands executes successfully
+    """
+    ocp = OCP(namespace=constants.OPENSHIFT_MACHINE_API_NAMESPACE)
+    ocp.exec_oc_cmd(f'delete machineset {machine_set}')
+    return True
```
Comment (keesturam) [inline ocs_ci/ocs/machine.py]: return after validating that the machine set is deleted.

File: ocs_ci/templates/openshift-infra/machine-set.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      providerSpec:
+        value:
+          ami:
+            id: ami-04a79d8d7cfa540cc
+          apiVersion: awsproviderconfig.openshift.io/v1beta1
+          blockDevices:
+            - ebs:
+                iops: 0
+                volumeSize: 120
+                volumeType: gp2
```
Comment (SvetlanaRedHat) [inline ocs_ci/templates/openshift-infra/machine-set.yaml]: should not be ceph or rbd ?

Comment (prsurve) [reply ocs_ci/templates/openshift-infra/machine-set.yaml]: no, as it's going to be root disk for instance

File: ocs_ci/ocs/machine.py [hunk 4]
```diff
@@ -138,6 +140,113 @@ def delete_machine_and_check_state_of_new_spinned_machine(machine_name):
...
+            region = machine.get('spec').get('template').get('spec').get('providerSpec').get('value').get(
+                'placement').get('region')
+            aws_zone = machine.get('spec').get('template').get('spec').get('providerSpec').get('value').get(
+                'placement').get('availabilityZone')
+            cls_id = machine.get('spec').get('selector').get('matchLabels').get(
+                'machine.openshift.io/cluster-api-cluster')
+            ami_id = machine.get('spec').get('template').get('spec').get('providerSpec').get('value').get(
+                'ami').get('id')
+            if aws_zone == f"{region}{zone}":
+                machineset_yaml = templating.load_yaml(
```
Comment (Avilir) [inline ocs_ci/ocs/machine.py]: Can not find the 'else' for that if.

Comment (ramkiperiy) [reply ocs_ci/ocs/machine.py]: This `if` itself inside for loop for each existing machineset data. 
i.e. Below default machineset of OCS cluster.
```
rperiyas-cls-w76c9-worker-us-east-2a 1 1 1 1 144m
rperiyas-cls-w76c9-worker-us-east-2b 1 1 1 1 144m
rperiyas-cls-w76c9-worker-us-east-2c 1 1 1 1 144m
```
Custom machineset.yaml required params are gathered from particular zone matching existing machineset. The 'else' would be just for not matching zones which is not required.

Why we need particular machineset ?
The reason would be to fetch aws zone, region, ami image id and cluster name.

File: ocs_ci/ocs/machine.py (hunk 1 above)
Comment (ramkiperiy) [reply ocs_ci/ocs/machine.py]: Updated function with return of machineset name and added validation

File: ocs_ci/ocs/machine.py (hunk 2 above)
Comment (ramkiperiy) [reply ocs_ci/ocs/machine.py]: Added validation in function, resolving comment.

**Pull request merged**